src/ray/protobuf/ray_client.proto

```diff
@@ -229,7 +240,34 @@ message KVListResponse {
   repeated bytes keys = 1;
 }
 
+message InitRequest {
+  // job_config of ray.init
+  bytes job_config = 1;
+}
+
+message InitResponse {
+  enum Status {
+    OK = 0;
+    MISSING_URIS = 1;
+    INCOMPATIBLE_RUNTIME_ENV = 2;
+  }
+  Status status = 1;
+}
+
+message PrepRuntimeEnvRequest {
+  RuntimeEnv runtime_env = 1;
+}
+
+message PrepRuntimeEnvResponse {
+  bool ok = 1;
+  string msg = 2;
+}
```

Comment:
See above comment-- I don't think we need to add PrepRuntimeEnv. The client can directly check if the package exists.

Reply:
It's to let the server know all info is ready.

It's like this, 
  - client connect to the server with job config
  - server will return whether it's ready or not
  - client will upload things that's missing
  - client notifies the server it's ready

Could you give me more detail about why do you think it's not necessary here?

Reply:
The client can check directly right?

- client connect to the server with job config [or abort if incompatible runtime env]
- client uploads anything missing (check directly using kv api)
- client notifies server

Reply:
Got it. I'm OK with this. But we still need two new entries. 1) for init; 2) for notify server it's ready. (right now, it's called PrepRuntimeEnv). What I did here is that when doing init, the server will return the missing one and the client will upload it. I can update it to this if it's ok with you.

- Init(job_config)
- RuntimeEnvReady()

Reply:
Got it, this is because we can't access the internal KV before ray is initialized, correct?

Maybe we can pass the job config in the initial connection request, so we don't need to add a separate RPC for Init?

Reply:
Yes, to access the internal kv, we need to connect to the server.

Oh, can we do that? That'll be great.

Reply:
I tried to pass it through metadata, but it looks like it's not valid to do it. If I include job_config in metadata, basically I need to always include it in metadata, which is not what I want to do. 
```

test_runtime_env.py::test_single_node[env1] E0325 05:26:51.442738110   15535 call.cc:923]                validate_metadata: {"created":"@1616650011.442714402","description":"Illegal header value","file":"src/core/lib/surface/validate_metadata.cc","file_line":44,"offset"
:0,"raw_bytes":"80 03 63 72 61 79 2e 6a 6f 62 5f 63 6f 6e 66 69 67 0a 4a 6f 62 43 6f 6e 66 69 67 0a 71 00 29 81 71 01 7d 71 02 28 58 0a 00 00 00 77 6f 72 6b 65 72 5f 65 6e 76 71 03 7d 71 04 58 1c 00 00 00 6e 75 6d 5f 6a 61 76 61 5f 77 6f 72 6b 65 72 73 5f 70 65 72 5f 70
 72 6f 63 65 73 73 71 05 4b 01 58 0b 00 00 00 6a 76 6d 5f 6f 70 74 69 6f 6e 73 71 06 5d 71 07 58 10 00 00 00 63 6f 64 65 5f 73 65 61 72 63 68 5f 70 61 74 68 71 08 5d 71 09 58 0b 00 00 00 72 75 6e 74 69 6d 65 5f 65 6e 76 71 0a 7d 71 0b 58 0b 00 00 00 77 6f 72 6b 69 6e 67
 5f 64 69 72 71 0c 58 10 00 00 00 2f 74 6d 70 2f 74 6d 70 71 73 6e 77 62 7a 69 36 71 0d 73 58 0a 00 00 00 63 6c 69 65 6e 74 5f 6a 6f 62 71 0e 89 75 62 2e '..cray.job_config.JobConfig.q.).q.}q.(X....worker_envq.}q.X....num_java_workers_per_processq.K.X....jvm_optionsq.]q
.X....code_search_pathq.]q.X....runtime_envq.}q.X....working_dirq.X..../tmp/tmpqsnwbzi6q.sX....client_jobq..ub.'\u0000"}
Traceback (most recent call last):
  File "<stdin>", line 14, in <module>
  File "/mnt/yic/ray-w2/python/ray/util/client_connect.py", line 34, in connect
    ignore_version=ignore_version)
  File "/mnt/yic/ray-w2/python/ray/util/client/__init__.py", line 73, in connect
    connection_retries=connection_retries)
  File "/mnt/yic/ray-w2/python/ray/util/client/worker.py", line 115, in __init__
    service_ready = bool(self.ping_server())
  File "/mnt/yic/ray-w2/python/ray/util/client/worker.py", line 434, in ping_server
    ray_client_pb2.ClusterInfoType.IS_INITIALIZED)
  File "/mnt/yic/ray-w2/python/ray/util/client/worker.py", line 385, in get_cluster_info
    resp = self.server.ClusterInfo(req, metadata=self.metadata)
  File "/mnt/yic/.local/lib/python3.6/site-packages/grpc/_channel.py", line 922, in __call__
    wait_for_ready, compression)
  File "/mnt/yic/.local/lib/python3.6/site-packages/grpc/_channel.py", line 909, in _blocking
    ),), self._context)
  File "src/python/grpcio/grpc/_cython/_cygrpc/channel.pyx.pxi", line 498, in grpc._cython.cygrpc.Channel.segregated_call
  File "src/python/grpcio/grpc/_cython/_cygrpc/channel.pyx.pxi", line 366, in grpc._cython.cygrpc._segregated_call
  File "src/python/grpcio/grpc/_cython/_cygrpc/channel.pyx.pxi", line 360, in grpc._cython.cygrpc._segregated_call
  File "src/python/grpcio/grpc/_cython/_cygrpc/channel.pyx.pxi", line 218, in grpc._cython.cygrpc._call
  File "src/python/grpcio/grpc/_cython/_cygrpc/channel.pyx.pxi", line 253, in grpc._cython.cygrpc._call
  File "src/python/grpcio/grpc/_cython/_cygrpc/channel.pyx.pxi", line 61, in grpc._cython.cygrpc._raise_call_error
ValueError: metadata was invalid: [('job_config', b'\x80\x03cray.job_config\nJobConfig\nq\x00)\x81q\x01}q\x02(X\n\x00\x00\x00worker_envq\x03}q\x04X\x1c\x00\x00\x00num_java_workers_per_processq\x05K\x01X\x0b\x00\x00\x00jvm_optionsq\x06]q\x07X\x10\x00\x00\x00code_search_p
athq\x08]q\tX\x0b\x00\x00\x00runtime_envq\n}q\x0bX\x0b\x00\x00\x00working_dirq\x0cX\x10\x00\x00\x00/tmp/tmpqsnwbzi6q\rsX\n\x00\x00\x00client_jobq\x0e\x89ub.')]
```

Do you have some idea of doing it without triggering this error?